### PR TITLE
Add Array#overlap?

### PR DESCRIPTION
**Pull request by fcheung**
This is semantically equivalent to `!(ary1 & ary2).empty?`, but more efficient and makes the intent more obvious

For example bundler [checks](https://github.com/ruby/ruby/blob/master/lib/bundler/runtime.rb#L61) whether the list of requested groups and the list of groups for a dependency has any overlap - it doesn't care what the overlap is, as long as it is non empty

In my personal projects we've found this to be a bottleneck when the arrays are large and where intersections are likely - `!(ary1 & ary2).empty?` keeps searching for all of the intersection even after we've found the first one, & creates extra garbage because of the intermediate array.

See also https://bugs.ruby-lang.org/issues/15976

### CODE REVIEW

File: array.c [hunk 1]
```diff
@@ -4631,6 +4631,56 @@ rb_ary_and(VALUE ary1, VALUE ary2)
     return ary3;
 }
 
+/*
+ *  call-seq:
+ *     ary.overlaps?(ary2)   -> true or false
+ *
+ *  Returns +true+ if the intersection of +ary+ and +ary2+ is non empty, else returns false. 
+ *
+ *  This is equivalent to (ary & ary2).empty?
```
Comment (nobu) [inline array.c]: Isn't it `!(ary & ary2).empty?` ?

Comment (fcheung) [reply array.c]: Of course - I meant to type `.any?`, as in the description here.

**Comment by MSP-Greg** [comment]
Maybe consider `overlap?` instead of `overlaps?`?  Similar to `File.exist?`...

**Comment by fcheung** [comment]
Thanks, that make sense - Range#cover?, Array#include? etc.

**Comment by nobu** [comment]
`any?` is not `!empty?`.
`[nil].any?` is false.

**Comment by fcheung** [comment]
Thanks, I had overlooked that, and it hadn’t come up in my usage. I will update accordingly

**Comment by Maumagnaguagno** [comment]
Related to Add Array#intersect? #1972

**Comment by travisofthenorth** [comment]
This is basically identical to https://github.com/ruby/ruby/pull/1972. 1972 has an additional performance improvement to avoid worst case and iterate over the shorter of the two arrays.

Apologies for the duplicated effort here. I could have included a bit more context in the PR description.

**Comment by Maumagnaguagno** [comment]
This PR can be closed since #1972 got merged.

**Pull request closed without merging**